apps/src/lib/client/tendermint_rpc_types.rs

```diff
@@ -34,65 +35,69 @@ pub struct TxResponse {
-                        "Couldn't find tx with hash {tx_hash} in events \
-                         {events:?}",
-                    );
-                    safe_exit(1)
-                }
-            };
+impl TryFrom<Event> for TxResponse {
+    type Error = String;
+
+    fn try_from(event: Event) -> Result<Self, Self::Error> {
+        fn missing_field_err(field: &str) -> String {
+            format!("Field \"{field}\" not present in event")
         }
-        let (index, _) = events
-            .get(&format!("{evt_key}.hash"))
-            .unwrap_or_else(tx_error!())
-            .iter()
-            .enumerate()
-            .find(|(_, hash)| hash == &tx_hash)
-            .unwrap_or_else(tx_error!());
-        let info = events[&format!("{evt_key}.info")][index].clone();
-        let log = events[&format!("{evt_key}.log")][index].clone();
-        let height = events[&format!("{evt_key}.height")][index].clone();
-        let code = events[&format!("{evt_key}.code")][index].clone();
-        let gas_used = events[&format!("{evt_key}.gas_used")][index].clone();
-        let initialized_accounts = events
-            [&format!("{evt_key}.initialized_accounts")]
-            .get(index)
-            .as_ref()
-            .map(|initialized_accounts| {
-                serde_json::from_str(initialized_accounts).unwrap()
-            })
-            .unwrap_or_else(|| {
-                eprintln!(
-                    "Tendermint omitted one of the expected indices in events"
-                );
-                Vec::new()
-            });
-        TxResponse {
+
+        let hash = event
+            .get("hash")
+            .ok_or_else(|| missing_field_err("hash"))?
+            .clone();
+        let info = event
+            .get("info")
+            .ok_or_else(|| missing_field_err("info"))?
+            .clone();
+        let log = event
+            .get("log")
+            .ok_or_else(|| missing_field_err("log"))?
+            .clone();
+        let height = event
+            .get("height")
+            .ok_or_else(|| missing_field_err("height"))?
+            .clone();
+        let code = event
+            .get("code")
+            .ok_or_else(|| missing_field_err("code"))?
+            .clone();
+        let gas_used = event
+            .get("gas_used")
+            .ok_or_else(|| missing_field_err("gas_used"))?
+            .clone();
+        let initialized_accounts = event
+            .get("initialized_accounts")
+            .map(String::as_str)
+            // TODO: fix finalize block, to return initialized accounts,
+            // even when we reject a tx?
+            .or(Some("[]"))
+            // NOTE: at this point we only have `Some(vec)`, not `None`
+            .ok_or_else(|| unreachable!())
+            .and_then(|initialized_accounts| {
+                serde_json::from_str(initialized_accounts)
+                    .map_err(|err| format!("JSON decode error: {err}"))
+            })?;
```

Comment:
What do we expect `event.get("initialized_accounts")` to look like now? It could be useful to factor out and have a test specifically for this parsing workaround we're doing. Is this an issue with Tendermint?

Reply:
no it's an issue with our finalize block implementation. at first I thought it was tendermint's fault, but it's namada code. when we reject a tx, `initialized_accounts` is not set in the `Event` attributes, so we need to do this crap, which is a bit unfortunate.

Reply:
Ah right, I made an issue - https://github.com/anoma/namada/issues/643 , we can eventually fix it and simplify this